Inflow-outflow boundary condition

mukul1992 opened the pull request:
## Summary

Enabling the `mass_inflow_outflow` (MIO) BC which allows for both inflow and outflow cells in a boundary.

## Pull request type

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

Please check the type of change introduced:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Checklist

The following is included:

- [ ] new unit-test(s)
- [x] new regression test(s)
- [ ] documentation for new capability

This PR was tested by running:

- the unit tests
  - [ ] on GPU <!-- note the OS and compiler -->
  - [ ] on CPU <!-- note the OS and compiler -->
- the regression tests
  - [ ] on GPU <!-- note the OS and compiler -->
  - [x] on CPU (macOS M3, gcc-13)

## Additional background

This functionality would be required for coupling with ERF and was started after discussions with @gdeskos and @mchurchf, along with @asalmgren, as it was a desired functionality within AMR-Wind too. Additionally, @hgopalan has expressed interest, and @marchdf has helped along the way.

## Description

The changes required can be organized into six categories:
1. adding MIO as one of the BCs, mapping it to a custom math BC, and adding the mechanics required for enabling UDFs.
2. the new TwoLayer UDF for testing, currently it implements a simple two-layer x-direction flow in opposite directions as a test, but this UDF can be modified to test progressively more complex cases, finishing with a regression test case such as vortex initialization and advection across the boundary.
3. the custom Neumann function for filling the outflow boundary cells in `MassInflowOutflowBC.cpp`
4. adding the inflow and outflow conditions in the WENO upwinding functions to map to `ext_dir` and `foextrap` behavior respectively.
5. calling the `enforceSolvability` routine in AMReX-Hydro before the mac-projection, which adjusts the outflux to match with influx (will link to the corresponding PR in Hydro).
6. the cmake scripts and input files for testing, these can be reorganized and cleaned up ultimately into a proper regression test as development completes. The input file being used for testing is `input_inout`.

marchdf commented on amr-wind/boundary_conditions/BCInterface.H:
```diff
@@ -49,7 +49,7 @@ public:
 
     //! User-defined functions for Dirichlet-type boundaries
-    std::pair<const std::string, const std::string> get_dirichlet_udfs();
+    std::array<const std::string, 3> get_dirichlet_udfs();
```
let's use `amrex::Array` for consistency with the rest of the code.

marchdf commented on amr-wind/boundary_conditions/velocity_bcs.H:
```diff
@@ -40,6 +41,11 @@ void register_inflow_vel_dirichlet(
             mesh, time, InflowOp(field));
+    } else if (inflow_udf == "TwoLayer") {
+        using InflowOp = BCOpCreator<udf::TwoLayer, WallOp>;
+        amrex::Print() << "***** Registering TwoLayer UDF" << std::endl;
```
remove debug print statement

marchdf commented on amr-wind/convection/incflo_godunov_ppm.H:
```diff
@@ -18,28 +18,34 @@ AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void Godunov_trans_xbc(
-        if (bclo == BCType::ext_dir) {
+        if (bclo == BCType::ext_dir ||
+            (bclo == BCType::user_1 && velp >= 0.0)) {
+            //Print() << "trans_xbc_domlo_extdir triggered at " << i << " " << j << " " << k << " " << n
```
when we get close to merging, let's remove all these print statements.

marchdf commented on cmake/amr-wind-utils.cmake:
```diff
@@ -79,6 +79,7 @@ macro(init_amrex_hydro)
   else()
     set(CMAKE_PREFIX_PATH ${AMReX-Hydro_DIR} ${CMAKE_PREFIX_PATH})
     find_package(AMReX-Hydro CONFIG REQUIRED)
+    add_subdirectory(${AMReX-Hydro_DIR} build)
```
Is this needed? Feels strange

mukul1992 replied:
I needed it to compile with an external AMReX-Hydro, but it may just be my inexperience with CMake. I will look into it..

mukul1992 replied:
I have removed this line. Something needs to be updated probably in Hydro for AW to compile with an external Hydro.

marchdf commented on amr-wind/physics/udfs/TwoLayer.H:
```diff
@@ -0,0 +1,83 @@
+{
+    struct DeviceOp
+    {
+        // clang-format off
```
remove this block

marchdf commented on amr-wind/physics/udfs/TwoLayer.cpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace amr_wind::udf {
+
+TwoLayer::TwoLayer(const Field& /*fld*/)
```
clean up this cpp and H files from comments and clang format off stuff

marchdf commented on amr-wind/projection/incflo_apply_nodal_projection.cpp:
```diff
@@ -6,6 +6,9 @@
 #include "amr-wind/core/field_ops.H"
 #include "amr-wind/wind_energy/ABL.H"
 
+// for debugging
```
remove these later

marchdf reviewed (changes requested):
Made a first quick pass.

marchdf commented on bctest/input_inflow:
```diff
@@ -0,0 +1,71 @@
+# This is a 2D poiseuille flow when run to steady state
```
move these into `test/test_files` and then add to `test/CMakeLists.txt`

mukul1992 replied:
yes, makes sense

mukul1992 replied:
I have done this, though the test currently fails for some reason — I will be debugging that.

marchdf commented:
Most of my comments right now are just noting the obvious things to clean up before actually merging. Thank you for taking this task on!!

mbkuhn commented:
Speaking of merging, it looks like there are some complicated merge conflicts between this branch and main. It may be helpful to rebase your branch to upstream/main to remove some of those conflicts and get the unrelated parts of this branch up-to-date. Removing the merge conflicts will also allow the CI tests to run.

mukul1992 commented:
@marchdf Yes, I will certainly clean up all the debug statements and incorporate other changes you recommended. 

@mbkuhn Yes, I will do the rebase over the next week along with cleaning things up.

Thanks for the early feedback!

mukul1992 commented on amr-wind/boundary_conditions/MassInflowOutflowBC.cpp:
```diff
@@ -0,0 +1,75 @@
+                    amrex::bdryLo(bx, idim),
+                    [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                        if (vel(i-ib, j-jb, k-kb, idim) < 0) {
+                            for (int n = 0; n < ncomp; n++) {
```
I believe that wouldn't work because we want to check the if condition (for outflow) on the original boundary value of velocity. For instance, we would want to check the the x-velocity at the x-boundary for the outflow, and if it is outflow, it is overwritten by the first internal cell. But then in the next loop iteration which operates on the y-velocity, we longer have the original x-velocity to check for outflow. I guess that's why the FixedGradientBC also has a separate component loop. I naively merged it with the ParallelFor when I started but figured out the above issue.

mukul1992 commented on amr-wind/boundary_conditions/MassInflowOutflowBC.cpp:
```diff
@@ -0,0 +1,75 @@
+    const auto islow = m_ori.isLow();
+    const auto ishigh = m_ori.isHigh();
+    const int nlevels = m_field.repo().num_active_levels();
+    const bool ib = (idim == 0), jb = (idim == 1), kb = (idim == 2);
```
will do

mukul1992 replied:
I changed the logic to use IntVect now.

mukul1992 commented:
PS. I have been able to rebase this code with the latest changes in the `Exawind/amr-wind` repo. I have also updated the amrex submod in the index (let me know if I did it incorrectly). And have updated the BCType from `user_1` to `direction_dependent` in the code.
And finally, here's the [draft PR for AMReX-Hydro](https://github.com/AMReX-Fluids/AMReX-Hydro/pull/129) that includes the outflow-correction for enforcing solvability.

mukul1992 commented on amr-wind/projection/incflo_apply_nodal_projection.cpp:
```diff
@@ -311,9 +329,23 @@ void incflo::ApplyProjection(
         if (!proj_for_small_dt and !incremental) {
```
I wanted to confirm that this if condition is correct. (It was there before my changes.)

mbkuhn replied:
yes -- basically this is never false except during initial iterations

mbkuhn replied:
I guess I'm not completely sure if it's necessary for these parts, but it should be inconsequential.

mukul1992 replied:
okay I won't touch it

mukul1992 commented:
@asalmgren @marchdf @mbkuhn 
I have cleaned up the code and addressed prior comments (received via github or discussed verbally).

There are two sticking point: the solution not reaching steady state for my test and the ctest failing (which I can run manually).

But I think the code is ready to be reviewed while we debug these.

mbkuhn commented on amr-wind/convection/incflo_godunov_ppm.H:
```diff
@@ -18,17 +18,20 @@ AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE void Godunov_trans_xbc(
     amrex::Real& hi,
-    amrex::Real& /* uedge */,
+    amrex::Real& velm,
+    amrex::Real& velp,
```
I don't understand the use of velm and velp as separate arguments when they are always passed in as the same value in incflo_godunov_advection.cpp . Will that become clear later?

mukul1992 replied:
I think I did this because some weno-related functions needed to pass different values. Let me check with Ann if we need this and also to confirm if the rest of the code in this file is consistent with how she handles things in Hydro as per the latest changes.

mukul1992 replied:
I changed it back to not having separate arguments in https://github.com/Exawind/amr-wind/pull/1066/commits/db9143fc47981e9db6522ac62354721007dad7be
which depends on #1199.

mukul1992 replied:
no longer relevant as this code is being moved to Hydro

mbkuhn commented on amr-wind/convection/incflo_godunov_weno.H:
I think this file is leftover from before we removed it from main. you'll have to git rm it

mukul1992 replied:
removed

mbkuhn commented on amr-wind/diffusion/incflo_diffusion.cpp:
```diff
@@ -20,6 +20,7 @@ Vector<Array<LinOpBCType, AMREX_SPACEDIM>> get_diffuse_tensor_bc(
             switch (bc) {
             case BC::pressure_inflow:
             case BC::pressure_outflow:
+            case BC::mass_inflow_outflow:
```
I'm guessing you had to pick Neumann out of the two options, for the sake of the possibility of outflow -- is there any concern of inaccuracy when the inflow is active?

mukul1992 replied:
I think we should be doing Dirichlet as we have the "correct" velocities in the boundary registers for both inflow and outflow, let me confirm.

mbkuhn replied:
Ok good, I'll let you confirm. I wasn't sure how you were filling boundaries when they switch to outflow. Yeah, Dirichlet would be best.

mukul1992 replied:
Changed diffusion BCs to Dirichlet: https://github.com/Exawind/amr-wind/pull/1066/commits/e18ba0b213bd9ca9138c53a3f68017243959e9e5

mukul1992 commented on amr-wind/equation_systems/icns/icns_advection.cpp:
```diff
@@ -266,6 +279,10 @@ void MacProjOp::operator()(const FieldState fstate, const amrex::Real dt)
         }
     }
 
+    if (m_has_inout_bndry) {
```
oops! You are very correct. I was using another strategy earlier for passing the internal flag but after changing it, it did not occur to me that I can access it directly from the advection operator. Thanks for pointing this out and not letting me add clutter.

I have updated it now.

mukul1992 commented on amr-wind/boundary_conditions/MassInflowOutflowBC.cpp:
```diff
@@ -0,0 +1,73 @@
+{
+    const auto& repo = m_field.repo();
+    const auto& velocity = repo.get_field("velocity");
+    // const auto bcvals = m_field.bc_values_device();
```
removed

mukul1992 commented on amr-wind/boundary_conditions/MassInflowOutflowBC.cpp:
```diff
@@ -0,0 +1,73 @@
+    const auto& velocity = repo.get_field("velocity");
+    // const auto bcvals = m_field.bc_values_device();
+    const int ncomp = m_field.num_comp();
+    // const int idx = static_cast<int>(m_ori);
```
removed

mukul1992 commented:
I was able to correct the issue that was causing my in-out BC ctest to fail in https://github.com/Exawind/amr-wind/pull/1066/commits/5856a85c184734ca346315a8044174fe5c5f2945.
While looking at this, I noticed a new issue related to the MAC velocity valid cell fills at the outflow for this BC, which did not exist earlier. I'll be working on that next week.

marchdf commented on amr-wind/core/FieldFillPatchOps.H:
```diff
@@ -441,12 +443,28 @@ public:
 
                     const auto& marr = mfab[mfi].array();
+                    amrex::FArrayBox tmp_fab(bx, mfab.nComp());
+                    amrex::Elixir tmp_eli = tmp_fab.elixir();
```
```suggestion
```

marchdf commented on amr-wind/core/FieldFillPatchOps.H:
```diff
@@ -441,12 +443,28 @@ public:
                     }
 
                     const auto& marr = mfab[mfi].array();
+                    amrex::FArrayBox tmp_fab(bx, mfab.nComp());
```
```suggestion
                    amrex::FArrayBox tmp_fab(bx, mfab.nComp(), amrex::The_Async_Arena());
```

marchdf reviewed (changes requested):
Just some minor code suggestions you can accept. Also, can you add the new BC type to the docs? Around here: https://github.com/Exawind/amr-wind/blob/main/docs/sphinx/user/inputs_Boundary_conditions.rst?plain=1#L12 and then below the existing mass_inflow BC you can add more details.
